### DISCUSSION TITLE
How can I get access to @app in a different file from main.py?

### DISCUSSION THREAD

**Discussion by renan-alonkin** [fastapi/fastapi#7982] [Questions] [answered]
Hello there,

**Description**

I have a middleware in my application, but, it is gigantic, and I want to move it to a different file. I would like to know if it is possible to access `@app` in a file that is not my `main.py`.

My `main.py` looks something like this: 
```
import uvicorn
from fastapi import Depends, FastAPI, Header, HTTPException
from routers.supply_change import predict

app = FastAPI(title=title, description=description,
              version=version, docs_url=docs_url, redoc_url=redoc_url)

@app.middleware("http")
async def add_process_time_header(request: Request, call_next):
    # Do some stuff
    response = await call_next(request)
    # Do some stuff
    return response

app.include_router(predict.router)
```
but I want something like this:

- `main.py`

```
import uvicorn
from fastapi import Depends, FastAPI, Header, HTTPException
from routers.supply_change import predict

app = FastAPI(title=title, description=description,
              version=version, docs_url=docs_url, redoc_url=redoc_url)

app.include_router(predict.router)
```

- `middleware.py`

```
@app.middleware("http")
async def add_process_time_header(request: Request, call_next):
    # Do some stuff
    response = await call_next(request)
    # Do some stuff
    return response
```

**Comment by sm-Fifteen** [comment, marked as answer]
You usually want to do it the other way around: Define a middleware function or class in a separate file, and then import that in your main where it will be attached to your app.

The decorator syntax for middlewares is equivalent to [`app.add_middleware(BaseHTTPMiddleware, dispatch=add_process_time_header)`](https://www.starlette.io/middleware/#basehttpmiddleware), so what you should be doing is:

```py
import uvicorn
from fastapi import Depends, FastAPI, Header, HTTPException
from routers.supply_change import predict

from starlette.middleware import BaseHTTPMiddleware
from .middleware import add_process_time_header

app = FastAPI(title=title, description=description,
              version=version, docs_url=docs_url, redoc_url=redoc_url)

app.add_middleware(BaseHTTPMiddleware, dispatch=add_process_time_header)

app.include_router(predict.router)
```

**Comment by renan-alonkin** [comment]
I did what you said, and it looks sooooo clean, thanks!

**Comment by tiangolo** [comment]
Thanks for the help here @sm-Fifteen ! :clap: :bow: 

Thanks for reporting back and closing the issue @RenanAlonkin  :+1:

**Comment by BertCatsburg** [comment]
Maybe a newer version of FastAPI (0.115.6) is the cause or I'm still doing something wrong, but I get a Typing error. 

1. The import of BaseHTTPMiddleware seems to have moved:
```
from starlette.middleware.base import BaseHTTPMiddleware
```

2. The add_middleware call gives a typing error: 
Code: 
```
app.add_middleware(BaseHTTPMiddleware, dispatch=APIKeyMiddleware)
```

IDE (Pycharm) Warning :
```
Expected type '_MiddlewareFactory[ParamSpec("P")]', got 'Type[BaseHTTPMiddleware]' instead 
```
